src/networking.c

```diff
@@ -2678,7 +2678,7 @@ NULL
         if (getLongLongFromObjectOrReply(c,c->argv[2],&id,NULL)
             != C_OK) return;
         struct client *target = lookupClientByID(id);
-        if (target && target->flags & CLIENT_BLOCKED) {
+        if (target && target->flags & CLIENT_BLOCKED && moduleBlockedClientMayTimeout(target)) {
```

Comment:
what happens if `moduleBlockedClientMayTimeout` returns 0 but `unblock_error` is 1? shouldn't we still reply to the blocked client with -UNBLOCKED?

Reply:
@guybe7 I think not, the way I see it a module that doesn't set a timeout callback doesn't expect spontaneous unblocking of clients.

Reply:
ok so you basically disabled CLIENT UNBLOCK for clients that don't expect a spontaneous unblock.. which means the only possibility to unblock the client is in the hands of the module.. but, isn't that the purpose of CLIENT UNBLOCK? shouldn't it do the best it can (even if the blocked client isn't "expecting" to be unblocked) in order to unblock a client?

Reply:
@guybe7 My POV is that referring to a command is blocking is mostly relevant for Redis built-in commands. Modules can use the blocking API to implement commands that run in background threads, and will use the blocking API, but from a user's perspective this is not really a blocking command.

Reply:
i think the important distinction is that the client side can expect to get unblocked with error, we care less about that. but we're afraid to throw the module off balance and into a problematic state, right?

so i agree we don't wanna corrupt the state for modules, but what if a certain module wants to support CLIENT unblock, but doesn't want timeout? i.e. what if he didn't define a timeout callback since it has nothing to do there, but will in some way gracefully handle a disconnected client? (they should expect client disconnection anyway, right?)

Reply:
@yossigo what about Oran's question?

maybe what we're saying is: if you want to be CLIENT UNBLOCKed you must define a timeout_callback. if you don't actually want a timeout, just pass `timeout=0`

btw what if i pass `timeout!=0` without a timeout callback? what happens it's timedout?

Reply:
Exactly that, if you want `CLIENT UNBLOCK` you need a timeout function - essentially preserving the current behavior. BTW going there is not very comfortable because you can't pass private data to the timeout function - but that was already discussed in the past and is a known shortcoming of the API.

Passing a non-zero timeout and a `NULL` callback will crash. Ways to address that:
1. Refuse the blocking call and return a `NULL` `RedisModuleBlockedClient`. That could potentially lead to other kinds of crashes because until now a `NULL` was never returned.
2. Silently skip the call. This would leave the client without a reply causing command/reply de-sync.
3. Produce our own custom error reply in this case. I think that makes most sense. WDUT?